url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/data/SampleTools.java

```diff
@@ -0,0 +1,352 @@
+        //
+        normalize(set, mean, stddev, idxs);
+    }
+    
+    /**
+     * Normalizes all features determined by a list of indices of a set 
+     * of samples. The normalization is based the given means and standard
+     * deviations. The method requires the samples to have the same input size
+     * which must also fit the length of the array of means and the array of 
+     * standard deviations. 
+     * <br></br>
+     * @param set The set of samples with the same input size.
+     * @param idxs A given list of indices determining the feature
+     * which are to be normalized.
+     */
+    public static void normalize(
+            final SampleSet set, 
+            final double[] mean, 
+            final double[] stddev,
+            final int ...idxs
+    ) {
+        if (set.size() == 0) return;
+        //
+        final int inputsize   = set.get(0).getInputSize();
+        //
+        for (Sample sample : set) {
+            //
+            final double[] input = sample.getInput();
+            //
+            int offset = 0;
+            for (int s = 0; s < sample.getInputLength(); s++) {
+                //
+                for (int i = 0; i < idxs.length; i++) {
+                    final int idx = idxs[i];
+                    final double x = input[offset + idx];
+                    input[offset + idx] = ((x - mean[idx]) / stddev[idx]);
+                }
+                //
+                offset += inputsize;
+            }
+        }
+    }
+    
+    private static int maxSize(final double[][] data) {
+        int max = 0;
+        for (double[] s : data) {
+            final int size = s.length;
+            if (size > max) max = size;
+        }
+        return max;
+    }
+    
+    private static double[][] transform(final String[] vector) {
+        double[][] data = new double[vector.length][];
+        for (int i = 0; i < vector.length; i++) {
+            String   v  = vector[i].trim();
+            String[] vs = v.split(DEFAULT_VALUEDELIMITER);
+            data[i]     = new double[vs.length];
+            for (int j = 0; j < data[i].length; j++) {
+                //int k = j;
+                data[i][j] = Double.parseDouble(vs[j]);
+            }
+        }
+        return data;
+    }
+    
+    private static void map(final double[][] source, final double[] dest, final int size) {
+        int idx = 0;
+        for (int i = 0; i < source.length; i++) {
+            for (int j = 0; j < size; j++) {
+                if (j < source[i].length) {
+                    dest[idx++] = source[i][j];
+                } else {
+                    idx++;
+                }
+            }
+        }
+    }
+    
+    public static SampleSet readCSV(final String filename) throws IOException {
+        SampleSet set = new SampleSet();
+        readCSV(filename, set);
+        return set;
+    }
+    
+    public static void readCSV(final String filename, final SampleSet set) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.